File: clients/java/src/main/java/net/weweave/tubewarder/client/SendRequest.java
package net.weweave.tubewarder.client;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * A request to be sent to a Tubewarder server.
 * Please read the docs at tubewarder.readthedocs.io.
 * @see <a href="http://tubewarder.readthedocs.io/en/latest/">tubewarder.readthedocs.io</a>
 */
public class SendRequest implements Serializable {
    private String token;
    private String template;
    private String channel;
    private Address recipient;
    private List<KeyValue> model = new ArrayList<>();
    private List<Attachment> attachments = new ArrayList<>();
    private String keyword;
    private String details;
    private Boolean echo = false;

    public SendRequest() {

    }

    public SendRequest(String token) {
        setToken(token);
    }

    public String getToken() {
        return token;
    }

    public void setToken(String token) {
        this.token = token;
    }

    public String getTemplate() {
        return template;
    }

    public void setTemplate(String template) {
        this.template = template;
    }

    public String getChannel() {
        return channel;
    }

    public void setChannel(String channel) {
        this.channel = channel;
    }

    public Address getRecipient() {
        return recipient;
    }

    public void setRecipient(Address recipient) {
        this.recipient = recipient;
    }

    public List<KeyValue> getModel() {
        return model;
    }

    public void addModelParam(KeyValue kv) {
        this.model.add(kv);
    }

    public void setModelFromMap(Map<String, Object> map) {
        this.model = new ArrayList<>();
        for (Map.Entry<String, Object> e : map.entrySet()) {
            addModelParam(new KeyValue(e.getKey(), e.getValue()));
        }
    }

    public List<Attachment> getAttachments() {
        return attachments;
    }

    public void addAttachment(Attachment attachment) {
        this.attachments.add(attachment);
    }

    public String getKeyword() {
        return keyword;
    }

    public void setKeyword(String keyword) {
        this.keyword = keyword;
    }

    public String getDetails() {
        return details;
    }

    public void setDetails(String details) {
        this.details = details;
    }

    public Boolean getEcho() {
        return echo;
    }

    public void setEcho(Boolean echo) {
        this.echo = echo;
    }
}

File: clients/php/src/SendRequest.php
<?php
namespace Tubewarder;
require('Attachment.php');
require('Address.php');

class SendRequest {
    private $token;
    private $template;
    private $channel;
    private $recipient;
    private $model = array();
    private $attachments = array();
    private $keyword;
    private $details;
    private $echo = false;

    public function __construct($token = '') {
        $this->token = $token;
        $this->recipient = new Address();
    }

    /**
     * @return string
     */
    public function getToken() {
        return $this->token;
    }

    /**
     * @param string $token
     */
    public function setToken($token) {
        $this->token = $token;
    }

    /**
     * @return string
     */
    public function getTemplate() {
        return $this->template;
    }

    /**
     * @param string $template
     */
    public function setTemplate($template) {
        $this->template = $template;
    }

    /**
     * @return string
     */
    public function getChannel() {
        return $this->channel;
    }

    /**
     * @param string $channel
     */
    public function setChannel($channel) {
        $this->channel = $channel;
    }

    /**
     * @return Address
     */
    public function getRecipient() {
        return $this->recipient;
    }

    /**
     * @param Address $recipient
     */
    public function setRecipient(Address $recipient) {
        $this->recipient = $recipient;
    }

    /**
     * @return array
     */
    public function getModel() {
        return $this->model;
    }

    /**
     * @param string $key
     * @param mixed $value
     */
    public function addModelParam($key, $value) {
        $this->model[$key] = $value;
    }

    /**
     * @return array
     */
    public function getAttachments() {
        return $this->attachments;
    }

    /**
     * @param Attachment $attachment
     */
    public function addAttachment(Attachment $attachment){
        $this->attachments[] = $attachment;
    }

    /**
     * @return string
     */
    public function getKeyword() {
        return $this->keyword;
    }

    /**
     * @param string $keyword
     */
    public function setKeyword($keyword) {
        $this->keyword = $keyword;
    }

    /**
     * @return string
     */
    public function getDetails() {
        return $this->details;
    }

    /**
     * @param string $details
     */
    public function setDetails($details) {
        $this->details = $details;
    }

    /**
     * @return boolean
     */
    public function isEcho() {
        return $this->echo;
    }

    /**
     * @param boolean $echo
     */
    public function setEcho($echo) {
        $this->echo = $echo;
    }

    /**
     * @return array
     */
    public function getObject() {
        $attachmentsArray = array();
        foreach ($this->getAttachments() as $a) {
            $attachmentsArray[] = $a->getObject();
        }
        $o = array(
            'token' => $this->getToken(),
            'template' => $this->getTemplate(),
            'channel' => $this->getChannel(),
            'recipient' => $this->getRecipient()->getObject(),
            'model' => $this->getModel(),
            'attachments' => $attachmentsArray,
            'keyword' => $this->getKeyword(),
            'details' => $this->getDetails(),
            'echo' => $this->isEcho()
        );
        return $o;
    }
}
?>
